server/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -583,17 +619,34 @@ private void closeShard(String reason, ShardId sId, IndexShard indexShard, Store
                 // this logic is tricky, we want to close the engine so we rollback the changes done to it
                 // and close the shard so no operations are allowed to it
                 if (indexShard != null) {
-                    try {
-                        // only flush if we are closed (closed index or shutdown) and if we are not deleted
-                        final boolean flushEngine = deleted.get() == false && closed.get();
-                        indexShard.close(reason, flushEngine);
-                    } catch (Exception e) {
-                        logger.debug(() -> "[" + shardId + "] failed to close index shard", e);
-                        // ignore
-                    }
+                    // only flush if we are closed (closed index or shutdown) and if we are not deleted
+                    final boolean flushEngine = deleted.get() == false && closed.get();
+                    // if the store is still open, want to keep it open while closing the shard
+                    final var hasStoreRef = store != null && store.tryIncRef();
```

Comment:
Do we need the incref/decref of store?

I would think that if the shard has an engine, it holds the necessary ref. If it does not, no ref is needed.

or in other words, the `IndexShard` instance should be reponsible for the store refs it needs itself?

Reply:
That's true, but previously we held the store open until after calling `afterIndexShardClosed` (and there was even a comment about doing so). I don't think it's important to do this looking at the implementations but, again, trying not to change semantics unnecessarily here.